### PR TITLE
Closes #439: Scroll to top

### LINKED ISSUE #439: Reset scroll position when the sort order changes
Really minor issue.
STR:

1. Scroll down
2. Change the sort order from A-Z to "Recent"

Now I'm basically at a random point in time (not withstanding #416) so it would be good to reset the scroll position whenever the sort order changes.

### PR DESCRIPTION
**Pull request by fluffyemily**
Closes https://github.com/mozilla-lockbox/lockbox-ios/issues/439

Automatically scrolls the `tableView` in `ItemsListView` to the top after a new sort has been performed.

Sends a `ScrollAction` when the sort order is changed and then to performs the scroll to top when the `ScrollAction` is received. 

It feels like overkill, but this was the only way I could find that would ensure that the table data update had occurred before the scroll to top was performed. If the scroll to top is performed before the table data update is complete, then the top of the table changes after the scroll to top completes and so the table is not actually at the top! 

Actually, there is another way. It involved adding a very small delay (0.1secs) to the observer before the `scrollToTop` call was made. You can see this in Part 1. In all conscience I could not submit a PR where the solution was to add a random timing delay to make something work so I had to find another way. This was it. 

P.S. I know this needs tests. Still trying to figure out how to add them....

### CODE REVIEW

**Comment by joeyg** [comment]
I tested this in the simulator and it worked most of the time. Occasionally it did not go all the way to the top. Here is a video: https://drive.google.com/open?id=1QkgRxXeR7PV4FF2xFbG3A54y0kK8Vm9g

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 1]
```diff
@@ -129,6 +136,21 @@ class ItemListPresenter: BaseItemListPresenter {
...
+        
+        self.dispatcher.register
+            .filterByType(class: ScrollAction.self)
+            .subscribe({ evt in
+                guard let action = evt.element else { return }
+                switch action {
+                case .toTop:
+                    DispatchQueue.main.async {
+                        self.view?.scrollToTop()
+                    }
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: Stick this in the `ItemListView`.

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 2]
```diff
@@ -129,6 +136,21 @@ class ItemListPresenter: BaseItemListPresenter {
         self.setupSpinnerDisplay()
 
         let itemSortObservable = self.userDefaultStore.itemListSort
+        
+        itemSortObservable.bind(to: self.listSortedObserver).disposed(by: self.disposeBag)
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: I'm not an Rx expect, but is it ok to re-use the same `asObservable` object here? 

```swift
self.userDefaultStore.itemListSort
```

gives you a new `Observable` each time.

Comment (fluffyemily) [reply lockbox-ios/Presenter/ItemListPresenter.swift]: I am pretty sure it's OK to bind/drive the same observable more than once, but hopefully @sashei has a better idea of whether that is OK.

**Comment by fluffyemily** [comment]
@joeyg can you retest to see whether this new strategy fixes the scrolling issue? If not, can you provide more details as to the scenarios in which the scrolling doesn't go all the way to the top?

**Comment by fluffyemily** [comment]
@joeyg Can you give this one a go re: the sometimes failing scroll to top?

**Comment by joeyg** [comment]
This works great! thanks for all your work on this @fluffyemily!

File: lockbox-iosTests/ItemListPresenterSpec.swift [hunk 3]
```diff
@@ -563,8 +576,18 @@ class ItemListPresenterSpec: QuickSpec {
 
             describe("sortingButton") {
                 beforeEach {
-                    self.subject.onViewReady()
-                    self.view.fakeOnSortingButtonPressed.onNext(())
+                    let voidObservable = self.scheduler.createColdObservable([next(50, ())])
+
+                    voidObservable
+                            .bind(to: self.subject.sortingButtonObserver)
```
Comment (devinreams) [inline lockbox-iosTests/ItemListPresenterSpec.swift]: @fluffyemily I tried merging `master` in to get this branch up to the latest. But I can't tell how to reconcile this test.

The changes from 4 days ago at #762 mean that `self.subject.sortingButtonObserver` doesn't quite exist here anymore and I couldn't figure this out myself.

Comment (fluffyemily) [reply lockbox-iosTests/ItemListPresenterSpec.swift]: It should be just a case of keeping 

```
let itemSettingObservable: TestableObservable<Setting.ItemListSort> = self.scheduler.createColdObservable([next(50, Setting.ItemListSort.alphabetically)])
                    itemSettingObservable
                        .bind(to: self.subject.listSortedObserver)
                        .disposed(by: self.disposeBag)

                    self.scheduler.start()
```

and losing the other changes.

Comment (fluffyemily) [reply lockbox-iosTests/ItemListPresenterSpec.swift]: Actually, this change causes all of my tests to fail as the fake button doesn't cause the `ScrollAction` to be fired. I may have to rethink these tests.

**Comment by fluffyemily** [comment]
@devinreams I've just rebased against the upstream master and resolved the test issues. Hope this helps.

File: lockbox-ios/Presenter/ItemListPresenter.swift (hunk 2 above)
Comment (sashei) [reply lockbox-ios/Presenter/ItemListPresenter.swift]: Yes, we do that in several other places in the app.

**Comment by devinreams** [comment]
>I've just rebased against the upstream master and resolved the test issues.

@fluffyemily thanks! nearly there I think.

Now I see a build error:

```
    ✗ lockbox-iosTests/ItemListViewSpec.swift:[50,22] property does not override any property from its superclass
517
    »         override var listSortedObserver: AnyObserver<Setting.ItemListSort> {
```

**Comment by fluffyemily** [comment]
Damned Xcode. Why allow me to run the `ItemListPresenterSpec` test when the `ItemListViewSpec` doesn't compile? Fixed. Sorry.

**Comment by devinreams** [comment]
>Fixed. Sorry.

No problem! Looks like everything is back to passing! (even though buddybuild disagrees and crashes on at least 1 test per build)

We will get this merged in shortly. Thanks for your contribution!

**Comment by devinreams** [comment]
Merging despite the buddybuild failure here as tests _are_ passing, just flaking out and crashing. See #756

**Pull request merged**